typedData/typedData_test.go

```diff
@@ -0,0 +1,346 @@
+		TypeName     string
+		ExpectedHash string
+	}
+	testSet := []testSetType{
+		{
+			TypedData:    typedDataExamples["baseExample"],
+			TypeName:     "StarkNetDomain",
+			ExpectedHash: "0x1bfc207425a47a5dfa1a50a4f5241203f50624ca5fdf5e18755765416b8e288",
+		},
+		{
+			TypedData:    typedDataExamples["baseExample"],
+			TypeName:     "Mail",
+			ExpectedHash: "0x13d89452df9512bf750f539ba3001b945576243288137ddb6c788457d4b2f79",
+		},
+		{
+			TypedData:    typedDataExamples["example_baseTypes"],
+			TypeName:     "Example",
+			ExpectedHash: "0x1f94cd0be8b4097a41486170fdf09a4cd23aefbc74bb2344718562994c2c111",
+		},
+		{
+			TypedData:    typedDataExamples["example_presetTypes"],
+			TypeName:     "Example",
+			ExpectedHash: "0x1a25a8bb84b761090b1fadaebe762c4b679b0d8883d2bedda695ea340839a55",
+		},
+		{
+			TypedData:    typedDataExamples["session_MerkleTree"],
+			TypeName:     "Session",
+			ExpectedHash: "0x1aa0e1c56b45cf06a54534fa1707c54e520b842feb21d03b7deddb6f1e340c",
+		},
+	}
+	for _, test := range testSet {
+		hash, err := test.TypedData.GetTypeHash(test.TypeName)
+		require.NoError(t, err)
+
+		require.Equal(t, test.ExpectedHash, hash.String())
+	}
+}
+
+// TestEncodeType tests the EncodeType function.
+//
+// It creates a mock typed data and calls the EncodeType method with the
+// type name. It checks if the returned encoding matches the expected
+// encoding. If there is an error during the encoding process, it fails the
+// test.
+//
+// Parameters:
+// - t: The testing.T object used for reporting test failures and logging test output
+// Returns:
+//
+//	none
+func TestEncodeType(t *testing.T) {
+	type testSetType struct {
+		TypedData      TypedData
+		TypeName       string
+		ExpectedEncode string
+	}
+	testSet := []testSetType{
+		{
+			TypedData:      typedDataExamples["baseExample"],
+			TypeName:       "StarkNetDomain",
+			ExpectedEncode: "StarkNetDomain(name:felt,version:felt,chainId:felt)",
+		},
+		{
+			TypedData:      typedDataExamples["baseExample"],
+			TypeName:       "Mail",
+			ExpectedEncode: "Mail(from:Person,to:Person,contents:felt)Person(name:felt,wallet:felt)",
+		},
+		{
+			TypedData:      typedDataExamples["example_array"],
+			TypeName:       "StarknetDomain",
+			ExpectedEncode: `"StarknetDomain"("name":"shortstring","version":"shortstring","chainId":"shortstring","revision":"shortstring")`,
+		},
+		{
+			TypedData:      typedDataExamples["example_baseTypes"],
+			TypeName:       "Example",
+			ExpectedEncode: `"Example"("n0":"felt","n1":"bool","n2":"string","n3":"selector","n4":"u128","n5":"i128","n6":"ContractAddress","n7":"ClassHash","n8":"timestamp","n9":"shortstring")`,
+		},
+		{
+			TypedData:      typedDataExamples["example_presetTypes"],
+			TypeName:       "Example",
+			ExpectedEncode: `"Example"("n0":"TokenAmount","n1":"NftId")"NftId"("collection_address":"ContractAddress","token_id":"u256")"TokenAmount"("token_address":"ContractAddress","amount":"u256")"u256"("low":"u128","high":"u128")`,
```

Comment:
If a "non-basic" type is used, does it get explicitly included in the encoding?

Reply:
Yes. Ref: https://github.com/starknet-io/SNIPs/blob/main/SNIPS/snip-12.md#encode_type